File: leetcode_problems/typescript/src/lc111_min_depth_of_binary_tree.ts
function minDepthRec(node: TreeNode | null): number {
  if (node === null) {
    return 10 ** 7;
  }
  if (node.left === null && node.right === null) {
    return 1;
  }
  let left = minDepthRec(node.left);
  let right = minDepthRec(node.right);
  return Math.min(left, right) + 1;
}

function minDepth(root: TreeNode | null): number {
  if (root === null) {
    return 0;
  }
  return minDepthRec(root);
}

File: leetcode_problems/cpp/lc111_min_depth_of_binary_tree.cpp
#include <algorithm>
#include <limits>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int minDepthRec(TreeNode *node)
    {
        if (node == nullptr)
        {
            return std::numeric_limits<int>::max();
        }
        if (node->left == nullptr && node->right == nullptr)
        {
            return 1;
        }
        int left = minDepthRec(node->left);
        int right = minDepthRec(node->right);
        return std::min(left, right) + 1;
    }

    int minDepth(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }
        return minDepthRec(root);
    }
};
